### ISSUE TITLE
Container isolation using firecracker

### ISSUE THREAD

**Issue by asim** [m3o/dev#430]

**Comment by domwong** [comment]
Update: We have network isolation using network policies.

@chrusty did an eval of firecracker and it was too complex for our needs right now. 

@chrusty did you do an eval of gvisor too?

**Comment by chrusty** [comment]
@domwong yes, FireCracker seems a bit too invasive for us right now (it would require us to roll our own Kubernetes clusters).

However, GVisor does seem feasible... it seems like Google's answer to FireCracker, but could be a drop-in replacement for Docker. It looks like Scaleway may be [offering this as an option when building new nodepools with Kapsule](https://developers.scaleway.com/en/products/k8s/api/) (their managed K8S service which we're already using).

My recommendation is to see how far we can get with GVisor through ContainerD on Kapsule. If we can do it and still retain the use of a manage K8S service then I don't see any down side.

**Comment by chrusty** [comment]
I've just had a go at creating a new cluster with containerd. Unfortunately this isn't supported in the web UI, but it _is_ possible using their CLI ("scw"). Just add the `pools.0.container-runtime=containerd` option to the cluster create command.

`scw k8s cluster create name=cruft version=1.19.3 pools.0.size=3 pools.0.node-type=GP1_XS pools.0.name=default pools.0.container-runtime=containerd cni=cilium`

I'm told however that I don't have sufficient permisisons to actually run this, so I've been unable to test any further. But if it is indeed this easy then it is something we could try with very little change to our current workflows.

**Comment by chrusty** [comment]
OK I managed to get something running.

- Created a new cluster with the `containerd` runtime (as above, but with appropriate credentials this time)
- Added a runtimeclass:
```yaml
apiVersion: node.k8s.io/v1beta1
kind: RuntimeClass
metadata:
  name: gvisor
handler: runsc
```
- Configured a pod with runtimeClassName set to `gvisor`:
```yaml
spec:
  runtimeClassName: gvisor
```

**Comment by chrusty** [comment]
So far the only down side I can think of would be if we were trying to run docker-in-docker by exposing the docker socket (build service for example). I'm not sure if that would work, possibly it might not even be an issue.
